### PR TITLE
Remove `BroadcastChannel` and cleanup annotations

### PR DESCRIPTION
**Pull request by pm47**
No more `ObsoleteCoroutinesApi`!

Best reviewed commit by commit.

### CODE REVIEW

**Review by t-bast — COMMENTED** [review]
LGTM, but I don't know enough about kotlin co-routines to give a proper ack, I think @sstone or @dpad85 should review those changes.

File: src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt
Comment (pm47) [reply src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt]: Removed in 261ce66.

File: src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt
Comment (pm47) [reply src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt]: Fixed in 261ce66.

File: src/commonMain/kotlin/fr/acinq/lightning/crypto/noise/Chacha20Poly1305CipherFunctions.kt
Comment (sstone) [inline src/commonMain/kotlin/fr/acinq/lightning/crypto/noise/Chacha20Poly1305CipherFunctions.kt:11]: Nice!

File: src/commonTest/kotlin/fr/acinq/lightning/tests/utils/LightningTestSuite.kt
Comment (sstone) [inline src/commonTest/kotlin/fr/acinq/lightning/tests/utils/LightningTestSuite.kt:15]: This looks suspicious: we he had one "guard" (`loggerIsSet`) per thread, now we have one global guard `LightningLoggerFactory` with no locks/guards to synchronize access to it from different threads ?

Comment (pm47) [reply src/commonTest/kotlin/fr/acinq/lightning/tests/utils/LightningTestSuite.kt:15]: Right. How about putting that code inside an `init` block if the companion object? Should be enough right? Since it's a singleton.

File: .github/workflows/test.yml
Comment (pm47) [inline .github/workflows/test.yml:67]: See the difference:
- before: https://github.com/ACINQ/lightning-kmp/runs/8012467463
- after: https://github.com/ACINQ/lightning-kmp/runs/8017194027

Latter build only shows failed test, much either to see what went wrong. Not sure what we gain with verbose logs for successful tests?

**Review by dpad85 — APPROVED** [review]
Tested successfully on Android and iOS. See https://github.com/ACINQ/phoenix/tree/use-cleanup

File: src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt
Comment (pm47) [inline src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt:199]: @sstone 

This follows Kotlin [migration's doc](https://github.com/Kotlin/kotlinx.coroutines/blob/d88a8c227a295f91251960c8fe6b21a02ddf704b/kotlinx-coroutines-core/common/src/flow/SharedFlow.kt#L100-L103):
> To migrate BroadcastChannel usage to SharedFlow, start by replacing usages of the BroadcastChannel(capacity) constructor with MutableSharedFlow(0, extraBufferCapacity=capacity) (broadcast channel does not replay values to new subscribers). Replace send and trySend calls with emit and tryEmit, and convert subscribers' code to flow operators.

The special negative capacity translated to `Int.MaxValue` in the [SharedFlow constructor](https://github.com/Kotlin/kotlinx.coroutines/blob/d88a8c227a295f91251960c8fe6b21a02ddf704b/kotlinx-coroutines-core/common/src/flow/SharedFlow.kt#L275) is a bit 😱 .

File: src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt
Comment (pm47) [inline src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt:155]: Same comment as https://github.com/ACINQ/lightning-kmp/pull/348#discussion_r956128122.

File: src/commonTest/kotlin/fr/acinq/lightning/tests/utils/LightningTestSuite.kt
Comment (pm47) [reply src/commonTest/kotlin/fr/acinq/lightning/tests/utils/LightningTestSuite.kt]: Done in [d3568d1](https://github.com/ACINQ/lightning-kmp/pull/348/commits/d3568d152381bc3444a08510496ec37cef1ace26).

Comment (pm47) [reply src/commonTest/kotlin/fr/acinq/lightning/tests/utils/LightningTestSuite.kt]: This was handled in #359.

File: .github/workflows/test.yml
Comment (pm47) [reply .github/workflows/test.yml]: This was done in #358

File: src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt
Comment (pm47) [reply src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumClient.kt]: So, the comment above wasn't accurate. There is a [check on non-negative value for `extraBufferCapacity`](https://github.com/Kotlin/kotlinx.coroutines/blob/d88a8c227a295f91251960c8fe6b21a02ddf704b/kotlinx-coroutines-core/common/src/flow/SharedFlow.kt#L270) in that constructor.

I settled with a non-zero value `extraBufferCapacity`, because doc says that the emitter will be blocked if there is no buffer and listeners are not ready. The buffer size `64` is the same as the default for [buffered channels](https://github.com/Kotlin/kotlinx.coroutines/blob/d88a8c227a295f91251960c8fe6b21a02ddf704b/kotlinx-coroutines-core/common/src/channels/Channel.kt#L732)

**Comment by sstone** [comment]
LGTM but we should build and test with Phoenix before we merge (this may be easier if this branch is renamed to `snapshot/cleanup`, github CI will then generate artefacts that we can publish to sonatype's snapshot repository)

**Comment by pm47** [comment]
> LGTM but we should build and test with Phoenix before we merge

@dpad85 can you test the branch? 🙏

**Comment by dpad85** [comment]
LGTM, tested on both Android and iOS 👍

**Review by sstone — APPROVED** [review]

**Pull request merged**
